miner/Miner.hs

```diff
@@ -212,36 +217,49 @@ mining e = do
     -- that successful mining. If `newWork` won the race instead, then the `go`
     -- call is automatically cancelled.
     --
-    miningSuccess :: BlockHeader -> IO ()
-    miningSuccess new = atomically $ modifyTVar' (results e) f
+    miningSuccess :: ChainId -> BlockHeight -> HeaderBytes -> IO ()
+    miningSuccess cid bh new = atomically $ modifyTVar' (results e) f
       where
-        key = T2 (_blockChainId new) (_blockHeight new)
+        key = T2 cid bh
         f m = M.insert key new . bool (prune m) m $ M.size m < int cap
 
     -- | Reduce the size of the result cache by half if we've crossed the "cap".
-    -- Clears old results out by `BlockCreationTime`.
+    -- Clears old results out by `BlockHeight`.
+    --
+    -- NOTE: There is an edge-case here involving hard-forks. Should a hard fork
+    -- occur and reset to a much lower `BlockHeight`, remote mining clients that
+    -- haven't been reset could potentially lose recent work when a `prune`
+    -- occurs.
     --
     prune :: ResultMap -> ResultMap
     prune = M.fromList
         . snd
-        . splitAt (int cap `div` 2)
-        . sortBy (compare `on` (_blockCreationTime . snd))
+        . splitAt (fromIntegral cap `div` 2)
+        . sortBy (compare `on` (\(T2 _ bh, _) -> bh))
         . M.toList
 
+    ver :: ChainwebVersion
+    ver = version $ args e
+
     -- | The maximum number of `BlockHeader`s to keep in the cache before
     -- pruning.
     --
     cap :: Natural
-    cap = 256 * (order . _chainGraph . version $ args e)
+    cap = 256 * (order $ _chainGraph ver)
 
     comp :: Comp
     comp = case cores $ args e of
         1 -> Seq
         n -> ParN n
 
-    -- This `head` should be safe, since `withScheduler` can only exit if it
+    -- TODO Use new `scheduler` to get safe head.
+    -- | This `head` should be safe, since `withScheduler` can only exit if it
     -- found some legal result.
-    go :: BlockHeader -> IO BlockHeader
-    go bh = fmap head . withScheduler comp $ \sch ->
-        replicateWork (int . cores $ args e) sch $
-            usePowHash (version $ args e) mine bh >>= terminateWith sch
+    go :: MWC.GenIO -> TargetBytes -> HeaderBytes -> IO HeaderBytes
+    go g tbytes hbytes = fmap head . withScheduler comp $ \sch ->
+        replicateWork (int . cores $ args e) sch $ do
+            -- TODO Be more clever about the Nonce that's picked to ensure that
+            -- there won't be any overlap?
+            n <- Nonce <$> MWC.uniform g
```

Comment:
The fix for a very dumb oversight. All the miners + their threads were starting work on the same nonce!

Reply:
:+1: